### ISSUE TITLE
Add option to turn off site hacks

### ISSUE THREAD

**Issue by szhu** [rsms/fb-mac-messenger#106]
Recently there have been a lot of issues created relating to elements being missing or present:
#96 ~~#97~~ #99 #101 #102 #104

These issues have existed for a few days without a fix. It is not beyond imagination that in the future Facebook might change the structure of the site again, which will again lead to a barrage of issues that leaves the app under-functional for a few days.

In light of all this, can there be an option to turn off `main.js`? I don't want to switch to using Chrome every time the app breaks, and this is an easy way to ensure that the app is always at least as good as the website.

**Comment by blakeembrey** [comment]
To be fair, even with `main.js` "turned off" it'd be the same experience. Everything you linked to was about the preferences cog, which should have been a more specific selector to start with and it would not have broken. The one thing not about the cog was about the beeping, which isn't a regression - without `main.js` there it would be beeping every time instead of a specific instance that was not accounted for. Not against this, just letting you know.

**Comment by szhu** [comment]
I removed issue 97, you're correct that it is not related.

Here are the issues:
- Menu items not working and settings cog missing*
- Search bar missing
- Slow typing

All these work in web browsers and should be fixed if `main.js` were turned off.

\* Turning off `main.js` won't fix menu items, but will allow the settings cog to consistently appear, so that I have a reliable way to access Settings.

**Comment by blakeembrey** [comment]
@szhu The settings cog is meant to be hidden though, that's the point. The issue is that the selector was really badly written and selected the first thing that matched. Since Facebook did an update, that selector changed and now a different element got selected. Those issues you've linked are all actually the same issue (not sure what's up with slow typing though). My comments here would be a good way to make sure this doesn't happen again https://github.com/rsms/fb-mac-messenger/pull/103/files#r29091931.

TL;DR: A badly written selector is used to hide the settings cog (on purpose), Facebook updated and the selector is matching and hiding the search bar instead (not hiding the cog).

Edit: Can't you just click the cog to access settings now? It works for me and that's what you'd do with `main.js` turned off.

**Comment by szhu** [comment]
Thanks for clarifying. Rewording my point in light of that:
- I want an option to always show the cog, in case Facebook updates the selector in the future.
- Search bar is still missing. Is that caused by the cog too?

**Comment by blakeembrey** [comment]
> Search bar is still missing. Is that caused by the cog too?

Yes, it's a broken selector. It was meant to target the cog and hide it, but now it targets the search bar and hides that instead.

> I want an option to always show the cog, in case Facebook updates the selector in the future.

If we use good enough selectors, I don't think it should happen like this. I like the cog hidden, it looks broken right now (because it is). When it breaks you can still click it anyway. If it can somehow be incorporated into the layout properly then maybe, but it's the same problem again. If we move the cog, that selector might break one day and it'll be under the traffic lights again or you can try moving the traffic lights instead but it'll be a weird UX.

**Comment by szhu** [comment]
Oh I have never tried clicking it when hidden, I guess that's good enough. Thanks!

**Comment by blakeembrey** [comment]
> Oh I have never tried clicking it when hidden

I think I must have worded something badly, sorry! You can't click it when it's hidden, but you can right now because it's not hidden. E.g. When the selector breaks and the cog shows up, you can now just click it instead. However, since the selector broke the menu item won't work anymore.

**Comment by szhu** [comment]
Oh okay that makes sense. What about the search bar? I can neither see it nor click on it..

**Comment by blakeembrey** [comment]
> I can neither see it nor click on it..

Yes, so you know how the cog is normally hidden now? The code has a selector that would find that element and set it to invisible. The problem is, that selector now matches the search bar. So the search bar is being set to invisible instead and the cog is visible.

**Comment by szhu** [comment]
Okay, I see. Now I understand your linked comment from the commit, you suggested to double-check that the selected thing is actually named "Settings" and not something else, so it's very unlikely to break ungracefully in the future. That sounds great!

**Comment by rsms** [comment]
This is why we should never have started injecting JS...

**Comment by rsms** [comment]
44d6d75587f508266de52ebf97f8ead33aa8a20e allows setting "main.js/disable" to disable main.js

```
defaults write me.rsms.fbmessenger "main.js/disable" -bool YES
```

Then reload the app. To reverse the effect:

```
defaults delete me.rsms.fbmessenger "main.js/disable"
```
